File: mm-php/src/io/metamessage/core/SimpleValue.php
<?php

namespace io\metamessage\core;

class SimpleValue
{
    public const SIMPLE_NULL = 0;
    public const NULL_BOOL = 1;
    public const NULL_INT = 2;
    public const NULL_FLOAT = 3;
    public const NULL_STRING = 4;
    public const NULL_BYTES = 5;
    public const FALSE = 6;
    public const TRUE = 7;

    public const CODE = 8;
    public const MESSAGE = 9;
    public const DATA = 10;
    public const SUCCESS = 11;
    public const ERROR = 12;
    public const UNKNOWN = 13;

    public const PAGE = 14;
    public const LIMIT = 15;
    public const OFFSET = 16;
    public const TOTAL = 17;
    public const ID = 18;
    public const NAME = 19;
    public const DESCRIPTION = 20;
    public const TYPE = 21;
    public const VERSION = 22;
    public const STATUS = 23;
    public const URL = 24;
    public const CREATE_TIME = 25;
    public const UPDATE_TIME = 26;
    public const DELETE_TIME = 27;
    public const ACCOUNT = 28;
    public const TOKEN = 29;
    public const EXPIRE_TIME = 30;
    public const KEY = 31;
    public const VAL = 32;

    public static function nameOf(int $sv): string
    {
        return match ($sv) {
            self::SIMPLE_NULL => 'simple_null',
            self::NULL_BOOL => 'null_bool',
            self::NULL_INT => 'null_int',
            self::NULL_FLOAT => 'null_float',
            self::NULL_STRING => 'null_string',
            self::NULL_BYTES => 'null_bytes',
            self::FALSE => 'false',
            self::TRUE => 'true',
            self::CODE => 'code',
            self::MESSAGE => 'message',
            self::DATA => 'data',
            self::SUCCESS => 'success',
            self::ERROR => 'error',
            self::UNKNOWN => 'unknown',
            self::PAGE => 'page',
            self::LIMIT => 'limit',
            self::OFFSET => 'offset',
            self::TOTAL => 'total',
            self::ID => 'id',
            self::NAME => 'name',
            self::DESCRIPTION => 'description',
            self::TYPE => 'type',
            self::VERSION => 'version',
            self::STATUS => 'status',
            self::URL => 'url',
            self::CREATE_TIME => 'create_time',
            self::UPDATE_TIME => 'update_time',
            self::DELETE_TIME => 'delete_time',
            self::ACCOUNT => 'account',
            self::TOKEN => 'token',
            self::EXPIRE_TIME => 'expire_time',
            self::KEY => 'key',
            self::VAL => 'value',
            default => 'unknown_simple_' . $sv,
        };
    }
}

File: mm-cs/src/MetaMessage/Core/SimpleValue.cs
namespace MetaMessage.Core;

public static class SimpleValue
{
    public const int NULL = 0;
    public const int NULL_BOOL = 1;
    public const int NULL_INT = 2;
    public const int NULL_FLOAT = 3;
    public const int NULL_STRING = 4;
    public const int NULL_BYTES = 5;
    public const int FALSE = 6;
    public const int TRUE = 7;
    public const int CODE = 8;
    public const int MESSAGE = 9;
    public const int DATA = 10;
    public const int SUCCESS = 11;
    public const int ERROR = 12;
    public const int UNKNOWN = 13;
    public const int PAGE = 14;
    public const int LIMIT = 15;
    public const int OFFSET = 16;
    public const int TOTAL = 17;
    public const int ID = 18;
    public const int NAME = 19;
    public const int DESCRIPTION = 20;
    public const int TYPE = 21;
    public const int VERSION = 22;
    public const int STATUS = 23;
    public const int URL = 24;
    public const int CREATE_TIME = 25;
    public const int UPDATE_TIME = 26;
    public const int DELETE_TIME = 27;
    public const int ACCOUNT = 28;
    public const int TOKEN = 29;
    public const int EXPIRE_TIME = 30;
    public const int KEY = 31;

    private static readonly string[] _names = new string[]
    {
        "null", "null_bool", "null_int", "null_float", "null_string", "null_bytes",
        "false", "true",
        "code", "message", "data", "success", "error", "unknown",
        "page", "limit", "offset", "total",
        "id", "name", "description", "type", "version", "status",
        "url", "create_time", "update_time", "delete_time",
        "account", "token", "expire_time",
        "key"
    };

    private static readonly Dictionary<string, int> _valueMap = new(StringComparer.OrdinalIgnoreCase);

    static SimpleValue()
    {
        for (int i = 0; i < _names.Length; i++)
        {
            _valueMap[_names[i]] = i;
        }
    }

    public static string NameOf(int value)
    {
        if (value >= 0 && value < _names.Length)
        {
            return _names[value];
        }
        return "unknown";
    }

    public static int? NameToValue(string name)
    {
        if (_valueMap.TryGetValue(name, out int value))
        {
            return value;
        }
        return null;
    }
}